### PR TITLE
[REFACTOR] Some structure changes to our REST layer 

### PR DESCRIPTION
**Pull request by cardoso**
@RocketChat/ios

TODO:
- [x] Fix tests
- [x] Check for bugs in affected features

Main changes:

- Unify errored and succeeded callbacks into one completion callback (switch .resource, .error): Cleaner syntax and no need to duplicate code for clean ups.
- Use associated type constraints (swift 4) instead of extension constraints: Better code completion and lighter on the compiler. Less limitations of working with extensions.

### CODE REVIEW

File: Rocket.Chat/Controllers/Chat/MembersListViewController.swift [hunk 1]
```diff
@@ -46,21 +46,24 @@ class MembersListViewData {
             let request = SubscriptionMembersRequest(roomId: subscription.rid, type: subscription.type)
             let options = APIRequestOptions.paginated(count: pageSize, offset: currentPage*pageSize)
 
-            API.current()?.fetch(request, options: options, succeeded: { result in
-                self.showing += result.count ?? 0
-                self.total = result.total ?? 0
-                if let members = result.members {
-                    self.membersPages.append(members.compactMap { $0 })
+            API.current()?.fetch(request, options: options) { response in
```
Comment (rafaelks) [inline Rocket.Chat/Controllers/Chat/MembersListViewController.swift]: We should be using `[weak self]` here 🤔

File: Rocket.Chat/Controllers/Chat/MessagesListViewController.swift [hunk 2]
```diff
@@ -68,28 +68,34 @@ class MessagesListViewData {
...
-                    self.handleMessages(
-                        fetchingWith: result.fetchMessagesFromRealm,
-                        showing: result.count,
-                        total: result.total,
-                        completion: completion
-                    )
-                }, errored: { _ in
-                    Alert.defaultError.present()
-                })
+                API.current()?.fetch(request, options: options) { response in
```
Comment (rafaelks) [inline Rocket.Chat/Controllers/Chat/MessagesListViewController.swift]: [weak self] here too

File: Rocket.Chat/Controllers/Chat/MessagesListViewController.swift [hunk 3]
```diff
@@ -68,28 +68,34 @@ class MessagesListViewData {
...
```
Comment (rafaelks) [inline Rocket.Chat/Controllers/Chat/MessagesListViewController.swift]: [weak self]

File: Rocket.Chat/Controllers/Subscriptions/NewRoomViewController.swift [hunk 4]
```diff
@@ -167,30 +167,33 @@ class NewRoomViewController: BaseViewController {
...
-        API.current()?.fetch(request, succeeded: { result in
-            guard
-                let success = result.success,
-                let name = result.name, success == true,
-                let auth = AuthManager.isAuthenticated()
-            else {
-                completion(false, result.error)
-                return
-            }
+        API.current()?.fetch(request) { response in
```
Comment (rafaelks) [inline Rocket.Chat/Controllers/Subscriptions/NewRoomViewController.swift]: [weak self]

Comment (cardoso) [reply Rocket.Chat/Controllers/Subscriptions/NewRoomViewController.swift]: self is not used inside this closure, so it would generate a warning

**Comment by rafaelks** [comment]
@cardoso Could you please fix the conflicts?

**Pull request merged**